### PR TITLE
[Bug][Colocate] when adding a table to the colocate group, we should check that the number of buckets per partition is the same

### PR DESCRIPTION
**Pull request by qzsee**
## Proposed changes

for example

```sql
CREATE TABLE `colocate_a` (
 dt date,
 k1 int,
 v1 int
) ENGINE=OLAP
DUPLICATE KEY(`k1`)
PARTITION BY RANGE(`dt`)
(PARTITION p1 VALUES [('2022-10-02'), ('2022-10-03'))
DISTRIBUTED BY HASH(`k1`) BUCKETS 2
PROPERTIES (
"replication_num" = "3",
"in_memory" = "false",
"storage_format" = "V2"
);

ALTER TABLE colocate_a set ("colocate_with" = "ab");

CREATE TABLE `colocate_b` (
 dt date,
 k1 int,
 v1 int
) ENGINE=OLAP
DUPLICATE KEY(`k1`)
PARTITION BY RANGE(`dt`)
(PARTITION p1 VALUES [('2022-10-02'), ('2022-10-03'))
DISTRIBUTED BY HASH(`k1`) BUCKETS 2
PROPERTIES (
"replication_num" = "3",
"in_memory" = "false",
"storage_format" = "V2"
);

ALTER TABLE colocate_b ADD PARTITION p2 VALUES [("2022-10-03"),("2022-10-04")) DISTRIBUTED BY HASH(k1) BUCKETS 10;

ALTER TABLE colocate_b set ("colocate_with" = "ab");
```

table `colocate_b` partition p2 set bucket num is 10 then take it into group `ab`.

In `ColocateTableCheckerAndBalancer` matchGroup occur :
```
java.lang.IllegalStateException: 2 vs. 10
303861     at com.google.common.base.Preconditions.checkState(Preconditions.java:508) ~[guava-30.0-jre.jar:?]
303862     at org.apache.doris.clone.ColocateTableCheckerAndBalancer.matchGroup(ColocateTableCheckerAndBalancer.java:242) ~[doris-fe.jar:1.2-SNAPSHOT]
303863     at org.apache.doris.clone.ColocateTableCheckerAndBalancer.runAfterCatalogReady(ColocateTableCheckerAndBalancer.java:95) ~[doris-fe.jar:1.2-SNAPSHOT]
303864     at org.apache.doris.common.util.MasterDaemon.runOneCycle(MasterDaemon.java:58) ~[doris-fe.jar:1.2-SNAPSHOT]
303865     at org.apache.doris.common.util.Daemon.run(Daemon.java:116) ~[doris-fe.jar:1.2-SNAPSHOT]
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by qzsee** [comment]
rebuild all

**Comment by qzsee** [comment]
run buildall

**Comment by qzsee** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 1]
```diff
@@ -4042,7 +4042,7 @@ public void modifyTableColocate(Database db, OlapTable table, String assignedGro
                 }
             } else {
                 // set to an already exist colocate group, check if this table can be added to this group.
-                groupSchema.checkColocateSchema(table);
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:4045]: why not check all partitions in this case? 
And please add comment here , or other people will not know the reason and may delete it during refactoring.

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 2]
```diff
@@ -2138,7 +2138,7 @@ private void createOlapTable(Database db, CreateTableStmt stmt) throws UserExcep
                 ColocateGroupSchema groupSchema = Env.getCurrentColocateIndex().getGroupSchema(fullGroupName);
                 if (groupSchema != null) {
                     // group already exist, check if this table can be added to this group
-                    groupSchema.checkColocateSchema(olapTable);
+                    groupSchema.checkColocateSchema(olapTable, false);
                     groupSchema.checkDynamicPartition(properties, olapTable.getDefaultDistributionInfo());
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:2142]: and here, pls add comment.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java (hunk 1 above)
Comment (qzsee) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:4045]: We add a table with many partitions to the colocate group, and we need to check whether all partitions comply with the colocate group specification

Comment (qzsee) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:4045]: > why not check all partitions in this case? And please add comment here , or other people will not know the reason and may delete it during refactoring.

We add a table with many partitions to the colocate group, and we need to check whether all partitions comply with the colocate group specification

**Comment by qzsee** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 51.62 seconds
 stream load tsv:          502 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          67 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17162051850 Bytes

**Comment by qzsee** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 50.79 seconds
 stream load tsv:          504 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.6 seconds inserted 10000000 Rows, about 337K ops/s
 storage size: 17168270264 Bytes

**Review by yiguolei — APPROVED** [review]
lgtm

**Comment by yiguolei** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 54.73 seconds
 stream load tsv:          534 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          67 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17167709486 Bytes

**Comment by qzsee** [comment]
run p0

**Review by yujun777 — APPROVED** [review]
LGTM

**Comment by yiguolei** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.7 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         18 seconds loaded 2358488459 Bytes, about 124 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17169933316 Bytes

**Pull request merged**